exec/java-exec/src/main/java/org/apache/drill/exec/ExecConstants.java

```diff
@@ -440,7 +441,7 @@ private ExecConstants() {
    * DEFAULT: 2048 MB
    */
   public static final String MAX_QUERY_MEMORY_PER_NODE_KEY = "planner.memory.max_query_memory_per_node";
-  public static final LongValidator MAX_QUERY_MEMORY_PER_NODE = new RangeLongValidator(MAX_QUERY_MEMORY_PER_NODE_KEY, 1024 * 1024, Long.MAX_VALUE);
+  public static final LongValidator MAX_QUERY_MEMORY_PER_NODE = new RangeLongValidator(MAX_QUERY_MEMORY_PER_NODE_KEY, 1024 * 1024, DrillConfig.getMaxDirectMemory());
```

Comment:
There is an assumption that all Drillbits in a cluster are started with the same amount of direct memory. What if this is not the case (cluster is composed of different size boxes)?

Reply:
I have raised this point to the team and they were fine with the change as currently the Planner assumes a Drill cluster is made of homogeneous nodes.

Reply:
The current implementation enables the customer to set MAX_QUERY_MEMORY_PER_NODE to unreasonable values, such as 1 TB.  This has lead to out-of-memory issues in some of our testing.

Drill should validate user settings to a reasonable point.  This is an attempt to move in that direction.  There is more work needed here.

I think it is challenging for the planner to make the best optimizations if each node in the cluster is a different size, and they want to use each node to its best capabilities.  I.e. make some nodes work harder than others.

One work-around is for the customer to set this option on the node with the largest Direct Memory if they want to set the option higher than recommended.  This could cause instability on the nodes that have less Direct Memory available.

Reply:
To achieve the goal of avoiding OOM errors, it is necessary to use node with the lowest amount of memory available, but what node will be used to validate max query memory per node depends on where the setting is applied.

Reply:
Yes, this change is not perfect by any means.  It is just a step towards a better solution.

A better solution is to have a global Resource Manager determine the right setting, where the Resource Manager knows what each node is capable of.  But we don't have that yet.

Reply:
@vrozov 

- The current code was not performing any validation (the maximum value was Long.MAX_VALUE)
- Now we added validation with the caveat that we are assuming the Drillbit are configured homogeneously
- Our code assumes the same configuration (cores-per-node, memory-per-node, and so on)

I suggest the following:
- Let's proceed with this new validation since it is an improvement (albeit not perfect); from the get go I emphasized the scope of this JIRA is limited
- Create a new JIRA to validate that all Drillbits are homogeneous; this will address your current feedback and many other validation deficiencies